src/BiDirectionalLinkedListImp.h

```diff
@@ -0,0 +1,152 @@
+//
+// Created by shvmo on 02/05/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+
+#include <cassert>
+#include "BiDirectionalLinkedList.h"
+
+using std::make_shared;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::BiDirectionalLinkedList() = default;
+
+template<class Type>
+BiDirectionalLinkedList<Type>::~BiDirectionalLinkedList() = default;
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getFirst() const {
+    return this->first;
+}
+
+template<class Type>
+BiDirectionalNode<Type> *BiDirectionalLinkedList<Type>::getFirstPointer() const {
+    return this->first.get();
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::getLast() const {
+    return this->last;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
+    if (this->first == nullptr) {
+        this->first = make_shared<BiDirectionalNode<Type>>();
+
+        this->first->setNext(nullptr);
+        this->first->setPrevious(nullptr);
+        this->first->setValue(newValue);
+
+        this->last = this->first;
+    } else {
+        shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
+
+        newLast->setNext(nullptr);
+        newLast->setPrevious(last);
+        newLast->setValue(newValue);
+
+        last->setNext(newLast);
+
+        this->last = newLast;
+    }
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::findNodeByValue(const Type &valueToFind) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            return current;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::deleteNodeByValue(const Type &valueToFind) {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            remove(current);
+
+            break;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printList(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        os << current->getValue() << " ";
+    }
+
+    os << std::endl;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printListReverse(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->last; current != nullptr;
+         current = current->getPrevious().lock()) {
+        os << current->getValue() << " ";
```

Comment:
what will `getValue()` return if `current->getPrevious()` is `nullptr` or expired?

Reply:
It's okay. Smart ptrs lock know how to deal with nulls
This use case of smart ptrs is alright because we own the entire least and know to put null in a smart ptr. And not let it point to the old deal located calue

Reply:
You are right though that smart_ptrs can potentially point to expired pointers, however because we control every node we know it won't happen in THIS case.

Reply:
Of course if an outside user has access to the node pointers he can screw things up, but they shouldn't modify them in the first case. Only pass them around from function to function for efficiency sake.

Reply:
boi i read about `weak_ptr.lock()`.
what i'm saying is, what if `current->previous` is null? then you'll try to access the method `.get()` from a null pointer.
am i wrong? `nullptr.get()` is a thing?